File: name-score/name-score-computation/src/main/java/com/shrader/namescore/scoring/strategy/FirstNameScoreStrategy.java
package com.shrader.namescore.scoring.strategy;

import com.shrader.namescore.scoring.NameScoreStrategy;
import org.springframework.stereotype.Component;

import java.util.Comparator;
import java.util.List;


@Component
public class FirstNameScoreStrategy implements NameScoreStrategy {

    private static final char BASE = 'A';

    @Override
    public int score(List<String> names) {
        names.sort(Comparator.comparing(Object::toString));

        int result = 0;
        for (int i = 0; i < names.size(); i++) {
            String name = names.get(i);

            result += scoreName(name, i + 1);
        }
        return result;
    }

    private int scoreName(String name, int multiplier) {
        int result = 0;

        for (int i = 0; i < name.length(); i++) {
            int ord = name.charAt(i);
            result += ord - BASE + 1;
        }
        return result * multiplier;
    }
}

File: name-score/name-score-computation/src/main/kotlin/com/shrader/namescore/scoring/strategy/FirstNameScoreStrategy.kt
package com.shrader.namescore.scoring.strategy

import com.shrader.namescore.scoring.NameScoreStrategy
import org.springframework.stereotype.Component


@Component
class FirstNameScoreStrategy : NameScoreStrategy {

    companion object {
        private const val BASE = 'A'.toInt()
    }

    override fun score(names: List<String>): Int {
        if(names.isEmpty()) {
            return 0
        }

        var sortedNames = names.toMutableList()
        sortedNames.sort()

        var result = 0
        for (i in sortedNames.indices) {
            val name = sortedNames[i]

            result += scoreName(name, i + 1)
        }
        return result
    }

    private fun scoreName(name: String, multiplier: Int): Int {
        var result = 0

        for (character in name) {
            val value = character.toInt()
            result += value - BASE + 1
        }
        return result * multiplier
    }
}
